### ISSUE TITLE
Exception in coroutines machinery (ClassCastException) affecting only few android 5 users 

### ISSUE THREAD

**Issue by Dmitry-Borodin** [Kotlin/kotlinx.coroutines#2763]
Kotlin 1.5.10, coroutines 1.4.3 (checked changelog - looks like not fixed)
This is happening to 200 users (800 crashes) only on android 5 (I have at least 8000 android 5 users in last week.)
Not reproducible.

```
Fatal Exception: kotlinx.coroutines.m0: Fatal exception in coroutines machinery for CancellableContinuation(DispatchedContinuation[Dispatchers.Default, Continuation at package.Repository$updateCacheVariants$1.invokeSuspend(Repository.kt)@2b36bba8]){Cancelled}@a821e66. Please read KDoc to 'handleFatalException' method and report this incident to maintainers
       at kotlinx.coroutines.DispatchedTask.handleFatalException(DispatchedTask.java:144)
       at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.java:115)
       at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.java:571)
       at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.java:750)
       at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.java:678)
       at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.java:665)
Caused by java.lang.ClassCastException: kotlin.a0.k.a.c cannot be cast to kotlinx.coroutines.internal.e
       at kotlinx.coroutines.CoroutineDispatcher.releaseInterceptedContinuation(CoroutineDispatcher.java:104)
       at kotlin.coroutines.jvm.internal.ContinuationImpl.releaseIntercepted(ContinuationImpl.java:118)
       at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(BaseContinuationImpl.java:39)
       at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.java:104)
       at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.java:571)
       at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.java:750)
       at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.java:678)
       at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.java:665)
```
      
 While I'm investigating why stacktrace is obfuscated (Firebase should deobfuscate stacktraces automatically), reporting it now so you may have some feedback.
 
 The code in referenced repository is as following
 
```
 		GlobalScope.launch {
			try {
// generated retrofit suspend function and access property of response model
				val response: List<Variants> = retrofitService.getAllVariants(request).variants 
				cachedVariants.saveVariantsToCache(response)
			}
			catch (e: Exception) {
				logCouldntSaveCache()
			}
                 }
```

**Comment by qwwdfsad** [comment]
Could be the same as #2736. 
Could you please elaborate on how did you manage to connect the exception with the given snippet? 
If it is an Anroid issue, probably extracting launch body to a separate method (along with `try/catch`) may help

**Comment by Dmitry-Borodin** [comment]
> Could you please elaborate on how did you manage to connect the exception with the given snippet?

In error message there is a reference "Continuation at <....>.Repository$updateCacheVariants$1.invokeSuspend(Repository.kt)@2b36bba8]){Cancelled}@a821e66" - 
In this file in this method there is only one suspend call, Simplified version of which is provided.

> If it is an Anroid issue, probably extracting launch body to a separate method (along with `try/catch`) may help
Thank you will try it out.

**Comment by Dmitry-Borodin** [comment]
I can add that extracting launch body to a separate method doesn't help.

**Comment by Dmitry-Borodin** [comment]
Deobfuscated version
```
Fatal Exception: kotlinx.coroutines.CoroutinesInternalError: Fatal exception in coroutines machinery for CancellableContinuation(DispatchedContinuation[Dispatchers.Default, Continuation at package.Repository$updateCacheVariants$1.invokeSuspend(Repository.kt)@2b36bba8]){Cancelled}@a821e66. Please read KDoc to 'handleFatalException' method and report this incident to maintainers
       at kotlinx.coroutines.DispatchedTask.handleFatalException(DispatchedTask.java:144)
       at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.java:115)
       at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.java:571)
       at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.java:750)
       at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.java:678)
       at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.java:665)
Caused by java.lang.ClassCastException
kotlin.coroutines.jvm.internal.CompletedContinuation cannot be cast to kotlinx.coroutines.internal.DispatchedContinuation
kotlin.coroutines.jvm.internal.ContinuationImpl.releaseIntercepted (ContinuationImpl.java:118)
kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith (BaseContinuationImpl.java:39)
kotlinx.coroutines.DispatchedTask.run (DispatchedTask.java:104)
kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely (CoroutineScheduler.java:571)
kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask (CoroutineScheduler.java:750)
kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker (CoroutineScheduler.java:678)
kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run (CoroutineScheduler.java:665)
```

**Comment by qwwdfsad** [comment]
Could you please elaborate on the coroutines versions? We expect this bug to be fixed in version 1.5.2

**Comment by Dmitry-Borodin** [comment]
I use 1,5,0, 
Will update and see how it goes.

**Comment by Dmitry-Borodin** [comment]
This is not fixed in 1.5.1. 
I still see this issue rarely and only on Android 5 devices. Stacktrace changed through, probably proguard changed line numbers.

**Comment by qwwdfsad** [comment]
Could you please share the new stacktrace?

**Comment by lordcodes** [comment]
I can reproduce this issue reliably on a Pixel C emulator with Android 10 (API 29). It doesn't happen with same app code on my Pixel 2.

```
2021-08-25 16:08:53.652 E/AndroidRuntime: FATAL EXCEPTION: main
Process: com.getbusy.android.getbusy.dev, PID: 9185
kotlinx.coroutines.CoroutinesInternalError: Fatal exception in coroutines machinery for CancellableContinuation(DispatchedContinuation[Dispatchers.Main, Continuation at com.getbusy.android.getbusy.authentication.model.UserIdentityRepository.fetchLatestUserIdentity(UserIdentityRepository.kt:20)@852144a]){Cancelled}@85d5fbb. Please read KDoc to 'handleFatalException' method and report this incident to maintainers
        at kotlinx.coroutines.DispatchedTask.handleFatalException(DispatchedTask.kt:144)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:115)
        at android.os.Handler.handleCallback(Handler.java:883)
        at android.os.Handler.dispatchMessage(Handler.java:100)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7356)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
     Caused by: java.lang.ClassCastException: kotlin.coroutines.jvm.internal.CompletedContinuation cannot be cast to kotlinx.coroutines.internal.DispatchedContinuation
        at kotlinx.coroutines.CoroutineDispatcher.releaseInterceptedContinuation(CoroutineDispatcher.kt:107)
        at kotlin.coroutines.jvm.internal.ContinuationImpl.releaseIntercepted(ContinuationImpl.kt:118)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:39)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:104)
        at android.os.Handler.handleCallback(Handler.java:883) 
        at android.os.Handler.dispatchMessage(Handler.java:100) 
        at android.os.Looper.loop(Looper.java:214) 
        at android.app.ActivityThread.main(ActivityThread.java:7356) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930) 
```

The line of app code the stacktrace refers to is just calling a suspend function on a 'Retrofit' interface.

**Comment by qwwdfsad** [comment]
Could you please share your reproducer? It would significantly simplify the investigation

**Comment by Dmitry-Borodin** [comment]
New stacktrace as asked for coroutines 1.5.1

```
Fatal Exception: kotlinx.coroutines.CoroutinesInternalError
Fatal exception in coroutines machinery for CancellableContinuation(DispatchedContinuation[Dispatchers.Default, Continuation at <my.package.class>.loadCacheVariants(Repo.kt)@a667b50]){Cancelled}@17151149. Please read KDoc to 'handleFatalException' method and report this incident to maintainers
Fatal Exception: kotlinx.coroutines.CoroutinesInternalError: Fatal exception in coroutines machinery for CancellableContinuation(DispatchedContinuation[Dispatchers.Default, Continuation at <my_package>.loadCacheVariants(Repo.kt)@a667b50]){Cancelled}@17151149. Please read KDoc to 'handleFatalException' method and report this incident to maintainers
       at kotlinx.coroutines.DispatchedTask.handleFatalException(DispatchedTask.java:144)
       at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.java:115)
       at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.java:571)
       at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.java:750)
       at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.java:678)
       at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.java:665)

Caused by java.lang.ClassCastException: kotlin.coroutines.jvm.internal.CompletedContinuation cannot be cast to kotlinx.coroutines.internal.DispatchedContinuation
       at kotlinx.coroutines.CoroutineDispatcher.releaseInterceptedContinuation(CoroutineDispatcher.java:107)
       at kotlin.coroutines.jvm.internal.ContinuationImpl.releaseIntercepted(ContinuationImpl.java:118)
       at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(BaseContinuationImpl.java:39)
       at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.java:104)
       at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.java:571)
       at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.java:750)
       at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.java:678)
       at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.java:665)
```

**Comment by Dmitry-Borodin** [comment]
Strange thing here is why google Crashlytics doesn't deobfuscate paths for
kotlinx.coroutines.CoroutinesInternalError
 kotlin.coroutines.jvm.internal.CompletedContinuation
 kotlinx.coroutines.internal.DispatchedContinuation
 I have to do it manually from proguard mapping file that was downloading to google together with the app. Other crashes are completely deobfuscated. I don't think it's connected with the issue, but just in case.

**Comment by LouisCAD** [comment]
@OliverO2 Can you comment in the YouTrack issue linked above? Ideally, you can also put a reproducing snippet there to help QA reproduce it.

**Comment by qwwdfsad** [comment]
Fixed in Kotlin 1.6.0

**Comment by brandlee** [comment]
my Kotlin version is 1.6.10, this bug happened again......
```
classpath 'com.android.tools.build:gradle:7.0.2'

classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:1.6.10"
```
